src/sentry/api/endpoints/project_rules.py

```diff
@@ -44,66 +48,219 @@ def pre_save_rule(instance, sender, *args, **kwargs):
+                and not current_rule_key_data
+                or not existing_rule.environment_id
+                and current_rule_key_data
+            ):
+                match_results.has_key = True
+
+        return match_results
+
+    def _actions_matcher(self, existing_rule: Rule, key_to_check: str) -> MatcherResult:
+        """
+        Special function that checks if the actions are the same against a rule.
+        """
+        match_results = MatcherResult()
+
+        existing_actions = existing_rule.data.get(key_to_check)
+        current_actions = self._rule_data.get(key_to_check)
+        if not existing_actions and not current_actions:
+            return match_results
+
+        # At this point, either both have the key, or one of the rules has the key, so this has to be true
+        match_results.has_key = True
+        # Only compare if both have the key
+        if existing_actions and current_actions:
+            match_results.key_matches = self._compare_lists_of_dicts(
+                keys_to_ignore=["uuid"], list1=existing_actions, list2=current_actions
+            )
+
+        return match_results
+
+    @classmethod
+    def _compare_lists_of_dicts(
+        cls,
+        keys_to_ignore: list[str],
+        list1: list[dict[Any, Any]] | None = None,
+        list2: list[dict[Any, Any]] | None = None,
+    ) -> bool:
+        if list1 is None or list2 is None:
+            return False
+
+        if len(list1) != len(list2):
+            return False
+
+        for i, left in enumerate(list1):
+            right = list2[i]
+            raw_left = {k: v for k, v in left.items() if k not in keys_to_ignore}
+            raw_right = {k: v for k, v in right.items() if k not in keys_to_ignore}
+
+            # TODO (Yash): This code commented below is the corrected logic which accounts for bad key values.
+            # clean_left = cls._get_clean_actions_dict(raw_left)
+            # clean_right = cls._get_clean_actions_dict(raw_right)
+            # if clean_left != clean_right:
+            #     return False
+            """
+            This is a bug in the current logic.
+            When comparing DB values to serialized values, the values that are `None` are not properly converted to
+            empty strings.
+            This means we end up incorrectly evaluating the actions aren't the same, when they actually are.
+            """
+            if raw_left != raw_right:
+                return False
+
+        return True
+
+    @classmethod
+    def _get_clean_actions_dict(cls, actions_dict: dict[any, any]) -> dict[any, any]:
+        """
+        Returns a dictionary where None is substituted with empty string to help compare DB values vs serialized values
+        """
+        cleaned_dict = {}
+        for k, v in actions_dict.items():
+            cleaned_dict[k] = "" if v is None else v
+
+        return cleaned_dict
+
+    def find_duplicate(self) -> Rule | None:
```

Comment:
My high-level feedback for this vs. the existing find_duplicate method is that this is too many levels too deep and hard to debug when something goes wrong.

Reply:
We'll figure out a way to simplify this in a follow up after we get some more insights from product